drivers: modem_cellular: Dynamic APN support

petrosyan-van opened the pull request:
This PR adds run‑time APN setting to the cellular driver, as well as possibility to register a callback on certain events (at the moment only for changes in modem_info) and the logic that programs the chosen APN into the modem before the normal dial sequence starts. Sample app demonstrates how to use dynamic APN setting via IMSI lookup.

#### Highlights

- **New API functions**
    - `cellular_set_callback(const struct device *dev, cellular_event_mask_t mask, cellular_event_cb_t cb, void *user_data)` -  subscribe to an event mask (e.g. CELLULAR_EVENT_MODEM_INFO_CHANGED).
    - `cellular_set_apn(const struct device *dev, const char *apn)` - inject or overwrite the APN at run time; automatically unblocks the driver if it is waiting in WAIT_FOR_APN.
* **Kconfig**
  * `MODEM_CELLULAR_APN` remains `internet` by default. If app sets it as empty, it will mean “wait for `cellular_set_apn()`”. (Making sure there is no regression)
- **Dynamic “set APN” chat script**
    - Introduces a new driver state **RUN_APN_SCRIPT** inserted between _open DLCI2_ and _run dial script_.
    - `modem_cellular_build_apn_script()` builds the chat script at run time using buffers that already exist in `struct modem_cellular_data`.
    - Every modem executes _AT+CGDCONT=1,"IP","<apn>"_. Sierra Wireless HL7800 additionally receives _AT+KCNXCFG=1,"GPRS","<apn>",,,"IPV4"_ (added via DT compat guard).
    - On script success we move to _run dial script_
    - On failure we continue repeating? 
- **Updated sample application**
  - Subscribe to MODEM_INFO_CHANGED,
  - Map IMSI → APN with the profile table,
  - Inject the APN at run time.
- **Testing**
    - Evaluated on a Quectel BG95 with Swisscom and Emnify SIMs; both boot to PPP with the correct APN.

I used this commit: https://github.com/zephyrproject-rtos/zephyr/commit/fdd89efd3ad7657f4f3b7478fab4104cc362cd4c from @hwilmers as inspiration.

My solution is not as elegant at all, as I couldn't find a really good way to handle the dynamic script build. i saw several discussions previously regarding this topic and some attempts to implement this feature, but for some reason the idea was dropped, perhaps due to the solution not being scalable. I found this discussion and PR here:

https://github.com/zephyrproject-rtos/zephyr/issues/67626
https://github.com/zephyrproject-rtos/zephyr/pull/67776

Nevertheless, I think this feature is important as it removes the need for per‑carrier firmware variants, which is a major practical gain. Feedback on improving scalability or restructuring the builder is highly appreciated :)

**Note**: Would be great to test the solution on other modems, especially on HL7800, which has 2 steps for APN script. Any help there would be great!

petrosyan-van committed:
drivers: modem_cellular: make APN configurable at run-time

* Implement the new cellular_set_apn() driver hook.
  The API validates the string, stores it in the driver’s
  private data and raises an internal event so the state
  machine can resume.

* Insert MODEM_CELLULAR_STATE_WAIT_FOR_APN between DLCI2 open
  and the APN-script step.
  – If CONFIG_MODEM_CELLULAR_APN is defined and non-empty the
    driver skips WAIT_FOR_APN (behaves exactly as before).
  – Otherwise the driver parks in WAIT_FOR_APN until the
    application calls cellular_set_apn().

* Add event-callback support (CELLULAR_EVENT_MODEM_INFO_CHANGED)
  so an app can subscribe to IMSI availability and kick its own
  APN selection logic.

Signed-off-by: Van Petrosyan <van.petrosyan@sensirion.com>

jeffwelder-ellenbytech commented:
This is really interesting!

A while ago I made some brute force edits to the `modem_cellular` state machine on a fork to switch between a primary and secondary APN with the `cellular.h` api to switch between a primary and secondary chat script.
(https://github.com/mpenate-ellenbytech/zephyr/commit/cd216aecab0bbdc937d1b73557fa6ceed2c6e523)

Couple other bg95 goodies I never got time to implement are here:
https://github.com/mpenate-ellenbytech/zephyr/commits/v3.7.99-eti?author=jeffwelder-ellenbytech

petrosyan-van commented:
> This is really interesting!
> 
> A while ago I made some brute force edits to the `modem_cellular` state machine on a fork to switch between a primary and secondary APN with the `cellular.h` api to switch between a primary and secondary chat script. ([mpenate-ellenbytech@cd216ae](https://github.com/mpenate-ellenbytech/zephyr/commit/cd216aecab0bbdc937d1b73557fa6ceed2c6e523))
> 
> Couple other bg95 goodies I never got time to implement are here: https://github.com/mpenate-ellenbytech/zephyr/commits/v3.7.99-eti?author=jeffwelder-ellenbytech

Thanks for the feedback! You have some interesting stuff there :)
Would be great if you can get some of your improvements merged at some point, especially the Network Tower info. This is something that I actually need in my project :D

bjarki-andreasen reviewed (changes requested):
Instead of making this a hardcoded part of a single cellular driver, I would highly recommend extending the cellular.h API with an entry like
```
int cellular_set_apn(const struct device *dev, const char *apn);
```
and allowing the application to store the list of APNs it knows of, either statically or dynamically (up to the application).
To know which APN to set, the application can already get IMSI itself using https://github.com/zephyrproject-rtos/zephyr/blob/64131856199342d0cb6507e5c57355c82f24dd69/include/zephyr/drivers/cellular.h#L215-L217 and do whatever lookup it needs :)

bjarki-andreasen commented:
> Thanks for the feedback :)
> 
> Yeah, make sense to move the responsibility of choosing the APN to the app layer, but this opens up several design questions and complicates the stat machine:
> 
>     * When the APN selection must be done by the app? To implement the dynamic look up on the app layer the driver should already pass the init stage and have IMSI ready, but then I assume it should not proceed to the dial state until APN is provided
> 
>     * Do we block the driver waiting for APN?

I think we will have to, we currently have little control over the modems states, and we probably should have an API specifically for registering and entering flight mode (control over CFUN basically) for the application to use. Right now, power on means register immediately, which is indeed making setting APN quite difficult to "insert" into the driver
> 
>     * Should driver notify app that IMSI is ready, so that app can figure out which APN to set?

Probably, adding a callback to get notifications when some `enum cellular_modem_info_type` is updated makes sense to me.
 
>     * What should happen if `cellular_set_apn()` is called after the connection has already established (don't know such use case, but what should be drivers behavior if app does such a call)

This would probably take affect next time the driver calls out, it could also return an error

>     * Should the current static APN be used as fallback, or APN logic should fully go out of scope of the driver? I am thinking for backward compatibility with existing solutions that rely on `CONFIG_MODEM_CELLULAR_APN`

Having the current static APN as a fallback is probably the way to go, until we can better control the states of the modem, I think we should move to force the app to set the APN in the future though

zafersn commented:
> > Instead of making this a hardcoded part of a single cellular driver, I would highly recommend extending the cellular.h API with an entry like
> > ```
> > int cellular_set_apn(const struct device *dev, const char *apn);
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > and allowing the application to store the list of APNs it knows of, either statically or dynamically (up to the application). To know which APN to set, the application can already get IMSI itself using
> > https://github.com/zephyrproject-rtos/zephyr/blob/64131856199342d0cb6507e5c57355c82f24dd69/include/zephyr/drivers/cellular.h#L215-L217
> > 
> > and do whatever lookup it needs :)
> 
> Thanks for the feedback :)
> 
> Yeah, make sense to move the responsibility of choosing the APN to the app layer, but this opens up several design questions and complicates the stat machine:
> 
> * When the APN selection must be done by the app? To implement the dynamic look up on the app layer the driver should already pass the init stage and have IMSI ready, but then I assume it should not proceed to the dial state until APN is provided
> * Do we block the driver waiting for APN?
> * Should driver notify app that IMSI is ready, so that app can figure out which APN to set?
> * What should happen if `cellular_set_apn()` is called after the connection has already established (don't know such use case, but what should be drivers behavior if app does such a call)
> * Should the current static APN be used as fallback, or APN logic should fully go out of scope of the driver? I am thinking for backward compatibility with existing solutions that rely on `CONFIG_MODEM_CELLULAR_APN`

Hi Petro,

I just happened to come across your PR — lucky timing! I wanted to say in because I had to tackle a very similar issue in my own PR for the HL78xx modem driver. The logic you're implementing is almost exactly what I ended up doing too. I’m not an expert or a core modem_cellular developer, but I’m trying to fully understand your use case, since there might be a scenario you’re covering that I haven’t accounted for in mine. Yeah, maybe my understanding is totally wrong :) anyway.

You mentioned:

> "When the APN selection must be done by the app? To implement the dynamic lookup on the app layer, the driver should already pass the init stage and have IMSI ready, but then I assume it should not proceed to the dial state until APN is provided."

What I’m trying to understand is: why does the driver need to block at that point?

From my perspective, since a SIM has just one IMSI and typically one APN, switching SIMs (and thus changing IMSI) usually means rebooting the modem anyway — at least on the HL7812 that’s the case. So, during board power-up, I associate the APN profile with the IMSI via Kconfig, then read the IMSI at the modem's init stage and match it to the correct APN. That way, I can parse and store the APN early and use it later before carrier_on.

To handle this, I just added a new internal state to configure GPRS (APN etc.) before reaching the carrier_on stage.

You also asked:

> "Should the driver notify the app that IMSI is ready so the app can set the APN? What if cellular_set_apn() is called after a connection is already established?"


In my case, if the app wants to change the APN after init, I just transition back into my hl78xx_on_enable_gprs_state_enter() state. In that state, I:

1. Put the modem in airplane mode
2. Set kcnx and cgdcont with the new APN
3. Bring it back to full functionality
4. Wait for network registration again; thats all before carrier_on.

So the modem driver itself never blocks, it just revisits GPRS config if needed. Works nicely without disrupting the init flow.

You also mentioned:

> "Should the current static APN be used as a fallback, or should APN logic be fully moved to the app? I’m thinking about backward compatibility with setups relying on CONFIG_MODEM_CELLULAR_APN."

Good point,  I had the same thought. Back when I was implementing it (on one of my lazier days 😄). But then decide it is not super important. I didn’t build in fallback logic. But yeah, if you misconfigure the APN, you’ll likely hit registration rejection, and probably with reject code 19 (ESM failure), if I remember right. So to support backwards compatibility, you’d probably need to detect that and treat it as a signal to retry or fall back to a default APN.

petrosyan-van commented:
@zafersn thanks for your message and for sharing your implementation; I’ll dig into it! 🙂

Regarding the blocking part: as @bjarki-andreasen mentioned the responsibility of setting the APN should go out from the driver scope and application should handle it, be it lookup with IMSI or any other method. Currently the driver does init script then directly goes to dial script and in general you call PM resume and driver handles everything by itself. If we are going to pass the APN via `cellular_set_apn()` function then the driver should start a dial script only when APN has been provided. If the APN is provided by the app prior even calling the PM resume then nothing will block as driver will have APN. So this is how the IMSI lookup would be in such case:

- Application registers a callback within a driver when IMSI is ready (or change in any modem info attribute)
- Application calls `pm_device_action_run(modem, PM_DEVICE_ACTION_RESUME);` -> Start the modem
- Driver starts the init script, and whenever IMSI is ready, calls the registered callback, moves to WAIT_FOR_APN state
- Application will do the IMSI lookup and figure out the APN and call `cellular_set_apn()` function
- Driver will ingest the provided APN and move to run APN script and then dial script

Btw by blocking I mean that driver just enters a state and waits for an event, and not real blocking. The key is not to proceed to dial script if there is no APN available

> From my perspective, since a SIM has just one IMSI and typically one APN, switching SIMs (and thus changing IMSI) usually means rebooting the modem anyway — at least on the HL7812 that’s the case. So, during board power-up, I associate the APN profile with the IMSI or ICCID via Kconfig, then read the IMSI at the modem's init stage and match it to the correct APN. That way, I can parse and store the APN early and use it later before carrier_on.

That’s essentially what my PR does now. The difference is that it was suggested to move the mapping logic to the application so the driver remains generic and only dials once it has a valid APN.

petrosyan-van committed:
drivers: cellular: introduce cellular_set_apn() API

Add a driver-level hook that lets applications provide (or override)
the APN string at run-time instead of hard-coding it in individual
drivers.

Signed-off-by: Van Petrosyan <van.petrosyan@sensirion.com>

petrosyan-van committed:
drivers: cellular: add event notification API

Introduces cellular_set_callback() and the first event
CELLULAR_EVENT_MODEM_INFO_CHANGED.  Drivers call the registered
callback whenever any modem-info field is updated.

Signed-off-by: Van Petrosyan <van.petrosyan@sensirion.com>

petrosyan-van committed:
samples: modem_cellular: add auto-APN detection and race-free DNS wait

This update makes the cellular-modem sample usable out-of-the-box on
SIMs that require different APNs, without rebuilding the firmware.

* Automatic APN selection
  * New Kconfig options
    * `CONFIG_SAMPLE_CELLULAR_MODEM_AUTO_APN` - enable logic
    * `CONFIG_SAMPLE_CELLULAR_APN_X` - APN for profile X
    * `CONFIG_SAMPLE_CELLULAR_IMSI_LIST_X` - IMSI list for profile X
  * The application now subscribes to the new
    `CELLULAR_EVENT_MODEM_INFO_CHANGED` event, waits for the IMSI, looks
    up the matching MCC+MNC and calls `cellular_set_apn()`.

* DNS-server detection
  * `NET_EVENT_DNS_SERVER_ADD` may fire together with
    `NET_EVENT_L4_CONNECTED`; the previous wait timed out.
  * Replaced the two consecutive waits with a single call that waits for
    both events (`DNS_SERVER_ADD | L4_CONNECTED`) in any order.

Signed-off-by: Van Petrosyan <van.petrosyan@sensirion.com>

rerickson1 commented:
@bjarki-andreasen, ready for your review.

bjarki-andreasen reviewed (changes requested):
Formatting changes should be avoided, please revert them, its just noise. Otherwise changes look promising :)

petrosyan-van commented on include/zephyr/drivers/cellular.h:
```diff
@@ -101,21 +127,30 @@ typedef int (*cellular_api_get_signal)(const struct device *dev,
+typedef int (*cellular_api_set_apn)(const struct device *dev, const char *apn);
+
+/** API for registering one asynchronous callback */
+typedef int (*cellular_api_subscribe_event)(const struct device *dev, enum cellular_event event,
```
then perhaps I should do it with `event_mask` instead of registering to a specific event to be more future proof. My idea was to have ability to register different callbacks for different events, but we can do it with one as well. Now instead of waiting for the `enum cellular_event event` it should be a mask of events I think

bjarki-andreasen reviewed (approved):
We really should start looking into moving some of these things into separate modules which can be individually tested. For now, this addition seems to provide quite a bit of value, but man, its a lot of code which is essentially untestable.

petrosyan-van commented:
> We really should start looking into moving some of these things into separate modules which can be individually tested. For now, this addition seems to provide quite a bit of value, but man, its a lot of code which is essentially untestable.

I agree, the modem driver is getting more and more complicated, and with more complex state machine. I am also not feeling 100% comfortable, as I only tested this with BG95 with 2 different SIM cards, but the code affects all other modems as well. If you already have an idea of a how we can improve the situation, I would be happy to help :)
